src/main/java/org/ethereum/beacon/discovery/database/ExpirationSet.java

```diff
@@ -0,0 +1,167 @@
+   */
+  public ExpirationSet(final long expirationDelayMillis, final Clock clock, final long size) {
+    this.expirationDelayMillis = expirationDelayMillis;
+    this.clock = clock;
+    if (size < 1) {
+      throw new RuntimeException("Minimal size is 1");
+    }
+    this.size = size;
+  }
+
+  /**
+   * Creates new set with records expiring in configured timeline after insertion
+   *
+   * @param expirationDelay Expiration delay, each record will be removed after this time
+   * @param clock Clock instance
+   * @param size Maximum size of a set, the oldest records will be removed to store new
+   */
+  public ExpirationSet(final Duration expirationDelay, final Clock clock, final long size) {
+    this(expirationDelay.toMillis(), clock, size);
+  }
+
+  private synchronized void clearExpired() {
+    final long currentTime = clock.millis();
+    boolean next = true;
+    while (next && !dataExpiration.isEmpty()) {
+      Pair<Long, V> last = dataExpiration.last();
+      if (last.getFirst() < currentTime) {
+        dataExpiration.remove(last);
+        data.remove(last.getSecond());
+      } else {
+        next = false;
+      }
+    }
+  }
+
+  @Override
+  public int size() {
+    clearExpired();
+    return data.size();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    clearExpired();
+    return data.isEmpty();
+  }
+
+  @Override
+  public boolean contains(Object o) {
+    clearExpired();
+    return data.contains(o);
+  }
+
+  @NotNull
+  @Override
+  public Iterator<V> iterator() {
+    clearExpired();
+    return data.iterator();
+  }
+
+  @NotNull
+  @Override
+  public Object[] toArray() {
+    clearExpired();
+    return data.toArray();
+  }
+
+  @NotNull
+  @Override
+  public <T> T[] toArray(@NotNull T[] a) {
+    throw new RuntimeException("Not implemented yet");
+  }
+
+  @Override
+  public synchronized boolean add(V v) {
+    clearExpired();
+    // no renewal
+    if (contains(v)) {
+      return false;
+    }
+    while (data.size() >= size) {
+      Pair<Long, V> last = dataExpiration.last();
+      dataExpiration.remove(last);
+      data.remove(last.getSecond());
+    }
+    dataExpiration.add(new Pair<>(clock.millis() + expirationDelayMillis, v));
+    data.add(v);
+    return true;
+  }
+
+  @Override
+  public synchronized boolean remove(Object o) {
+    throw new RuntimeException("Not implemented yet");
+  }
+
+  @Override
+  public boolean containsAll(@NotNull Collection<?> c) {
+    return data.containsAll(c);
+  }
+
+  @Override
+  public boolean addAll(@NotNull Collection<? extends V> c) {
+    clearExpired();
+    c.forEach(this::add);
+    return true;
+  }
+
+  @Override
+  public boolean retainAll(@NotNull Collection<?> c) {
+    throw new RuntimeException("Not implemented yet");
+  }
+
+  @Override
+  public boolean removeAll(@NotNull Collection<?> c) {
+    throw new RuntimeException("Not implemented yet");
+  }
```

Comment:
Rather than leaving methods unimplemented, I'd just not implement the generic `Set` interface as we don't need a generic data structure here, just the `add`, and `contains` methods with the automatic expiry.